src/main/java/org/roaringbitmap/longlong/RoaringTreeMap.java

```diff
@@ -0,0 +1,284 @@
+			}
+
+			// We get a 'select' query for a single bitmap: should fit in an int
+			final int givenBitmapSelect = (int) (j - previousBucketCardinality);
+
+			MutableRoaringBitmap bitmaps = linkedBitmaps.get(insertionPoint);
+
+			int low = bitmaps.select(givenBitmapSelect);
+
+			int high = sortedHighs[insertionPoint];
+
+			return pack(high, low);
+		}
+	}
+
+	public LongIterator iterator() {
+		final Iterator<Map.Entry<Integer, MutableRoaringBitmap>> it = hiToBitmap.entrySet().iterator();
+
+		return new LongIterator() {
+
+			protected int currentKey;
+			protected IntIterator currentIt;
+
+			@Override
+			public boolean hasNext() {
+				if (currentIt == null) {
+					// Were initially empty
+					if (!moveToNextEntry(it)) {
+						return false;
+					}
+				}
+
+				while (true) {
+					if (currentIt.hasNext()) {
+						return true;
+					} else {
+						if (!moveToNextEntry(it)) {
+							return false;
+						}
+					}
+				}
+			}
+
+			/**
+			 *
+			 * @param it
+			 * @return true if we MAY have more entries. false if there is definitely nothing more
+			 */
+			private boolean moveToNextEntry(Iterator<Map.Entry<Integer, MutableRoaringBitmap>> it) {
+				if (it.hasNext()) {
+					Map.Entry<Integer, MutableRoaringBitmap> next = it.next();
+					currentKey = next.getKey();
+					currentIt = next.getValue().getIntIterator();
+
+					// We may have more long
+					return true;
+				} else {
+					// We know there is nothing more
+					return false;
+				}
+			}
+
+			@Override
+			public long next() {
+				if (hasNext()) {
+					return pack(currentKey, currentIt.next());
+				} else {
+					throw new IllegalStateException("empty");
+				}
+			}
+
+			@Override
+			public LongIterator clone() {
+				throw new UnsupportedOperationException("TODO");
+			}
+		};
+	}
+
+	public long rankLong(long id) {
+		int x = (int) (id >> 32);
+		int y = (int) id;
+
+		ensureCumulatives(x);
+
+		int bitmapPosition = Arrays.binarySearch(sortedHighs, 0, sortedHighs.length, x);
+
+		if (bitmapPosition >= 0) {
+			// There is a bucket holding this item
+
+			final long previousBucketCardinality;
+			if (bitmapPosition == 0) {
+				previousBucketCardinality = 0;
+			} else {
+				previousBucketCardinality = sortedCumulatedCardinality[bitmapPosition - 1];
+			}
+
+			MutableRoaringBitmap bitmap = linkedBitmaps.get(bitmapPosition);
+
+			// Rank is previous cardinality plus rank in current bitmap
+			return previousBucketCardinality + bitmap.rankLong(y);
+		} else {
+			// There is no bucket holding this item: insertionPoint is previous bitmap
+			int insertionPoint = -bitmapPosition - 1;
+
+			if (insertionPoint == 0) {
+				// this key is before all inserted keys
+				return 0;
+			} else {
+				// The rank is the cardinality of this previous bitmap
+				return sortedCumulatedCardinality[insertionPoint - 1];
+			}
+		}
+	}
+
+	protected void ensureCumulatives(int x) {
+		// Check if missing data to handle this rank
+		if (!allValid && firstHiNotValid <= x) {
```

Comment:
I do not assume this. The idea of ensureCumulatives is to compute it lazily, and dismiss these values on any call to .add/.addLong. For now, any call to .add leads to the dismissing of any ensureCumulatives about the added long.

@lemire, would it be possible/relevant to return a boolean on RoaringBitmap.add, to indicate the Bitmap has been modified (like Collection.add)? It would help refining the dismissing strategy of ensureCumulatives